os/arch/arm/src/amebad/amebad_timer_lowerhalf.c

```diff
@@ -0,0 +1,348 @@
+
+#include <sys/types.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/timer.h>
+#include <tinyara/irq.h>
+#include "timer_api.h"
+
+#ifdef CONFIG_TIMER
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct amebad_gpt_lowerhalf_s {
+	const struct timer_ops_s *ops;	/* Lowerhalf operations */
+	gtimer_t obj;
+	uint8_t irq_id;
+	bool started;				/* True: Timer is started */
+	tccb_t callback;
+	void *arg;					/* Argument passed to upper half callback */
+	uint32_t gpt_timeout;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+static int amebad_gpt_start(struct timer_lowerhalf_s *lower);
+static int amebad_gpt_stop(struct timer_lowerhalf_s *lower);
+static int amebad_gpt_getstatus(struct timer_lowerhalf_s *lower, struct timer_status_s *status);
+static int amebad_gpt_settimeout(struct timer_lowerhalf_s *lower, uint32_t timeout);
+static void amebad_gpt_setcallback(struct timer_lowerhalf_s *lower, CODE tccb_t callback, void *arg);
+static int amebad_gpt_ioctl(struct timer_lowerhalf_s *lower, int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+static const struct timer_ops_s g_timer_ops = {
+	.start = amebad_gpt_start,
+	.stop = amebad_gpt_stop,
+	.getstatus = NULL,
+	.settimeout = amebad_gpt_settimeout,
+	.setcallback = amebad_gpt_setcallback,
+	.ioctl = NULL,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt0_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt1_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt2_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+static struct amebad_gpt_lowerhalf_s g_gpt3_lowerhalf = {
+	.ops = &g_timer_ops,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: amebad_gpt_handler
+ *
+ * Description:
+ *   timer interrupt handler
+ *
+ * Input Parameters:
+ *
+ * Returned Values:
+ *
+ ****************************************************************************/
+
+static void amebad_timout_handler(uint32_t data)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)data;
```

Comment:
we need null check for priv here.

Reply:
Sure, will add DEBUGASSERT(priv) for all the priv